lib/src/util/lazy.dart

```diff
@@ -0,0 +1,111 @@
+// limitations under the License.
+
+library over_react.lazy;
+
+import 'package:over_react/over_react.dart';
+import 'package:react/react.dart' as react;
+
+import '../component_declaration/function_component.dart';
+
+/// A HOC that creates a "lazy" component that lets you defer loading a component’s code until it is rendered for the first time.
+///
+/// Returns a `UiFactory` you can use just render in your tree. While the code for the lazy component is still loading,
+/// attempting to render it will suspend. Use <Suspense> to display a loading indicator while it’s loading.
+/// [load] is a function that should return a `Future<UiFactory<TProps>>` that resolves to the component to be rendered.
+/// [_config] should be a `UiFactoryConfig<TProps>` or `null` and is only `dynamic` to avoid an unnecessary cast in the boilerplate.
+///
+/// React will not call [load] until the first time the component is rendered.
+/// After React first calls [load], it will wait for it to resolve, and then render the resolved value.
+/// Both the returned Future and the Future's resolved value will be cached, so React will not call [load] more than once.
+/// If the Future rejects, React will throw the rejection reason for the nearest Error Boundary to handle.
+///
+/// Example:
+/// ```dart
+/// import 'package:over_react/over_react.dart';
+///
+/// part 'main.over_react.g.dart';
+///
+/// mixin ALazyComponentPropsMixin on UiProps {}
+///
+/// UiFactory<ALazyComponentPropsMixin> ALazyComponent = lazy(
+///   () async {
+///     final componentModule = await loadComponent();
+///     return uiForwardRef(
+///       (props, ref) {
+///         return (componentModule.AnotherComponent()
+///           ..ref = ref
+///           ..addProps(props)
+///         )(props.children);
+///       },
+///       _$ALazyComponentConfig,
+///     );
+///   },
+///   _$ALazyComponentConfig
+/// );
+/// ```
+///
+/// > __NOTE:__ A lazy component MUST be wrapped with a `Suspense` component to provide a fallback ui while it loads.
+///
+/// ```dart
+/// (Suspense()
+///   ..fallback = Dom.p()('Loading...')
+/// )(
+///    ALazyComponent()(),
+/// );
+/// ```
+/// See: <https://react.dev/reference/react/lazy>.
+UiFactory<TProps> lazy<TProps extends UiProps>(
+    Future<UiFactory<TProps>> Function() load, /* UiFactoryConfig<TProps> */ dynamic _config) {
+      _config ??= UiFactoryConfig();
+  if (_config is! UiFactoryConfig<TProps>) {
+    throw ArgumentError('_config is required when using a custom props class and should be a UiFactoryConfig<TProps>. Make sure you are '
+        r'using either the generated factory config (i.e. _$FooConfig) or manually declaring your config correctly.');
+  }
+  // ignore: invalid_use_of_protected_member
+  var propsFactory = _config.propsFactory;
+
+  final lazyFactoryProxy = react.lazy(() async {
+    final factory = await load();
+    return factory().componentFactory!;
+  });
```

Comment:
Aw one thing I just realized is that there's no more `displayName` logic, so the react-dart wrapper just shows up as `Anonymous`

![Screenshot 2024-10-04 at 3 05 54 PM](https://github.com/user-attachments/assets/4c904016-edb9-487c-96ce-e37fc7f9c83c)

I'd be down to not worry about that in this PR, though, and do that as a follow-up improvement. Your call

Reply:
I mean, i could just add like `}, displayName: 'lazy(anonymous)')` to the react-dart side for the time being lol

Reply:
or actually would `'lazy(${factory.type.displayName})'` work in the react-dart side?

Reply:
![Screenshot 2024-10-07 at 9 29 40 AM](https://github.com/user-attachments/assets/01bd7181-cfc6-4f79-8e76-be06ddf1924f)